File: 2nd Year/[LFA] PF - ANTLR Compiler for a new Robotics-Focused Programming Language/reference/croblink.cpp
/*
    This file is part of ciberRatoToolsSrc.

    Copyright (C) 2001-2011 Universidade de Aveiro

    ciberRatoToolsSrc is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    ciberRatoToolsSrc is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
// CRobLink implementation

#include "croblink.h"
#include "structureparser.h"

#include <iostream>

#ifdef CIBERQTAPP
#include <qsocketnotifier.h>
#endif
#include <qxml.h>

using namespace std;

#define MSGMAXSIZE (4096)


/*!
 * Composes register message and sends it to server.
 */
void CRobLink::send_register_message(char *rob_name,int rob_id)
{
    // register in server
	char xml[MSGMAXSIZE];
    const char fmt[] = "<Robot Name=\"%s\" Id=\"%d\"></Robot>";
	sprintf(xml, fmt, rob_name, rob_id);

    if(port.send_info(xml, strlen(xml)+1)!=true)
    {
        // cerr << "Failed Send Init" << endl;
		Status=-1;
		return;
    }
}

void CRobLink::send_register_message(char *rob_name,int rob_id, double irSensorAngles[])
{
    // register in server
	char xml[MSGMAXSIZE];

    int n = sprintf(xml, "<Robot Name=\"%s\" Id=\"%d\">",
                    rob_name, rob_id);

    for(int i=0; i < NUM_IR_SENSORS; i++)
        n += sprintf(xml+n,"<IRSensor Id=\"%d\" Angle=\"%g\"/>",
                     i, irSensorAngles[i]);

	sprintf(xml+n,"</Robot>");

    if(port.send_info(xml, strlen(xml)+1)!=true)
    {
        // cerr << "Failed Send Init" << endl;
		Status=-1;
		return;
    }
}

/*!
 * Composes register message and sends it to server.
 */
void CRobLink::send_robotbeacon_register_message(char *rob_name,int rob_id, double height)
{
    // register in server
	char xml[MSGMAXSIZE];
    const char fmt[] = "<RobotBeacon Name=\"%s\" Id=\"%d\" Height=\"%g\"/>";
	sprintf(xml, fmt, rob_name, rob_id, height);

    if(port.send_info(xml, strlen(xml)+1)!=true)
    {
        // cerr << "Failed Send Init" << endl;
		Status=-1;
		return;
    }
}

/*!
 * Waits for server reply and parses status, simulation parameters and
 * assigns the new UDP port to this robot.
 */
void CRobLink::parse_server_reply(void)
{
    // get reply confirming or denying registration 
    char xml[MSGMAXSIZE];
    int recv_ret = port.recv_info(xml, MSGMAXSIZE);
    if(recv_ret == -1)
    {
        cerr << "Failed Init confirmation" << recv_ret << endl;
		Status = -1;
		return;
    }
    //cerr << "XML=\"" << xml <<"\"\nXMLEND\n";

	/* set source of xml document */
	QXmlInputSource source;
	source.setData(QString(xml));

	/* set parser handler */
    StructureParser handler(simParam.nBeacons);

	/* parse xml document with handler */
    QXmlSimpleReader reader;
    reader.setContentHandler(&handler);
    if( !reader.parse(source) ) {
       Status=-1;
       return;
    }


    simParam = *(handler.getSimParam());
    simParam.showValues();

    port.SetRemote(port.GetLastSender());
}

CRobLink::CRobLink(char *rob_name, int rob_id, char *host) : measures(0), port(6000,host,0)
{
    Status = 0;

    if(!port.init())
	{
        // cerr << "Failed socket init" << endl;
		Status=-1;
		return;
    }

    send_register_message(rob_name, rob_id);
    if( Status != 0 ) return;

    parse_server_reply();
    if( Status != 0 ) return;

#ifdef CIBERQTAPP
    QSocketNotifier *notifier=new QSocketNotifier(port.socketfd,QSocketNotifier::Read);

    QObject::connect(notifier,SIGNAL(activated(int)),this,SIGNAL(NewMessage()));

    notifier->setEnabled(true);
#endif

    Status = 0;
}

CRobLink::CRobLink(char *rob_name, int rob_id, double irSensorAngles[], char *host) :  measures(0), port(6000,host,0)
{
    Status = 0;

    if(!port.init())
	{
        // cerr << "Failed socket init" << endl;
		Status=-1;
		return;
    }

    send_register_message(rob_name,rob_id,irSensorAngles);
    if( Status != 0 ) return;

    parse_server_reply();
    if( Status != 0 ) return;

#ifdef CIBERQTAPP
    QSocketNotifier *notifier=new QSocketNotifier(port.socketfd,QSocketNotifier::Read);

    QObject::connect(notifier,SIGNAL(activated(int)),this,SIGNAL(NewMessage()));

    notifier->setEnabled(true);
#endif

    Status = 0;
}

CRobLink::CRobLink(char *rob_name, int rob_id, double height, char *host) : measures(0), port(6000,host,0) 
{
    Status = 0;

    if(!port.init())
	{
        // cerr << "Failed socket init" << endl;
		Status=-1;
		return;
    }

    send_robotbeacon_register_message(rob_name, rob_id, height);
    if( Status != 0 ) return;

    parse_server_reply();
    if( Status != 0 ) return;

#ifdef CIBERQTAPP
    QSocketNotifier *notifier=new QSocketNotifier(port.socketfd,QSocketNotifier::Read);

    QObject::connect(notifier,SIGNAL(activated(int)),this,SIGNAL(NewMessage()));

    notifier->setEnabled(true);
#endif

    Status = 0;
}

CRobLink::~CRobLink()
{
}

int CRobLink::ReadSensors()
{
	char xml[4096];
    int n = port.recv_info(xml, 4096);
	if (n == -1) return n;

	//cerr << "ReadSensors: " << "\"" << xml << "\"";
	
	/* set source of xml document */
	QXmlInputSource source;
	source.setData(QString(xml));

	/* set parser handler */
    StructureParser handler(simParam.nBeacons);

    //cerr << "ReadSensors: nBeacons=" << simParam.nBeacons << "\n";

	/* parse xml document with handler */
    QXmlSimpleReader reader;
    reader.setContentHandler(&handler);
    reader.parse(source);

	///////////////////////////////////////////////////
	measures = *(handler.getMeasures());  
	//measures.showValues();
	///////////////////////////////////////////////////
	
//    for(unsigned int i=0; i<5;i++)
//       if (measures.hearMessage[i]!=QString())
//           printf("ReadSensors: Message From %d: \"%s\"\n", i, measures.hearMessage[i].latin1());

    return n;
}

void CRobLink::requestGround()
{
    char xml[128];
    const char fmt[] = "<Actions> <SensorRequests Ground=\"Yes\" /> </Actions>\n";
    unsigned int n = sprintf(xml,"%s", fmt);
    port.send_info(xml,n+1);
    //cout << xml;
}

void CRobLink::requestCompass()
{
    char xml[128];
    const char fmt[] = "<Actions> <SensorRequests Compass=\"Yes\" /> </Actions>\n";
    unsigned int n = sprintf(xml, "%s", fmt);
    port.send_info(xml,n+1);
    //cout << xml;
}

void CRobLink::requestBeacon(int id)
{
    char xml[128];
    const char fmt[] = "<Actions> <SensorRequests Beacon%d=\"Yes\" /> </Actions>\n";
    unsigned int n = sprintf(xml, fmt, id);
    port.send_info(xml,n+1);
    //cout << xml;
}

void CRobLink::requestObstacle(int id)
{
    char xml[128];
    const char fmt[] = "<Actions> <SensorRequests IRSensor%d=\"Yes\" /> </Actions>\n";
    unsigned int n = sprintf(xml, fmt, id);
    port.send_info(xml,n+1);
    //cout << xml;
}

void CRobLink::requestSensors(int nReqs, va_list ap)
{
    char *sensId;
    char xml[2048]="<Actions>\n\t<SensorRequests ";
	int  s,n;

	n=strlen(xml);

	for(s=0; s < nReqs; s++) {
	   sensId = va_arg(ap,char *);
	   n += sprintf(xml+n,"%s=\"Yes\" ",sensId);
	}

	n+= sprintf(xml+n,"/>\n</Actions>");

    port.send_info(xml,n+1);
}

void CRobLink::DriveMotors(double lPow,double rPow)
{
    char xml[1024];
    const char fmt[] = "<Actions LeftMotor=\"%g\" RightMotor=\"%g\"/>\n";
	//sprintf(xml, fmt, lPow*1000.0+2000.5, rPow*1000.0+2000.5);
	unsigned int n = sprintf(xml, fmt, lPow, rPow);
    port.send_info(xml,n+1);
	//cout << xml;
}

void CRobLink::Say(char *msg)
{
    char xml[1024];
    const char fmt[] = "<Actions><Say><![CDATA[%s]]></Say></Actions>\n";
    //char *fmt = "<Actions> <Say Ground=\"Yes\"/> </Actions>";
    unsigned int n = sprintf(xml, fmt, msg);
    port.send_info(xml,n+1);
	//cout << xml;
}

void CRobLink::SetReturningLed(bool val)
{
    char xml[128];
    const char fmt[] = "<Actions LeftMotor=\"%g\" RightMotor=\"%g\" ReturningLed=\"%s\"/>\n";
	//sprintf(xml, fmt, lPow*1000.0+2000.5, rPow*1000.0+2000.5, "Off");
	unsigned int n = sprintf(xml, fmt, 0.0, 0.0, (val?"On":"Off"));
    port.send_info(xml,n+1);
	//cout << xml;
}

void CRobLink::SetVisitingLed(bool val)
{
    char xml[128];
    const char fmt[] = "<Actions LeftMotor=\"%g\" RightMotor=\"%g\" VisitingLed=\"%s\"/>\n";
	//sprintf(xml, fmt, lPow*1000.0+2000.5, rPow*1000.0+2000.5, "Off");
	unsigned int n = sprintf(xml, fmt, 0.0, 0.0, (val?"On":"Off"));
    port.send_info(xml,n+1);
	//cout << xml;
}

void CRobLink::Finish(void)
{
	char xml[] = "<Actions LeftMotor=\"0.0\" RightMotor=\"0.0\" EndLed=\"On\"/>\n";
	unsigned int n = strlen(xml);
	//cout << xml;
    port.send_info(xml,n+1);
}

File: 5th Year/[RMI] Assignment 2 - Robotic Agent to solve Movement, Location, Mapping, and Planning problems on a maze/croblink.py

import socket

UDP_IP = "127.0.0.1"
UDP_PORT = 6000

NUM_IR_SENSORS = 4

class CRobLink:


    def __init__ (self, robName, robId, host):
        self.robName = robName
        self.robId = robId
        self.host = host

        self.sock = socket.socket(socket.AF_INET, # Internet
                             socket.SOCK_DGRAM) # UDP
        
        msg = '<Robot Id="'+str(robId)+'" Name="'+robName+'" />'
        
        self.sock.sendto(msg.encode(), (host, UDP_PORT))  # TODO consider host arg
        data, (host,self.port) = self.sock.recvfrom(1024)
        #print "received message:", data, " port ", self.port

        parser = sax.make_parser()
        
        # Tell it what handler to use
        handler = StructureHandler()
        parser.setContentHandler( handler )
        
        # Parse reply 

        d2 = data[:-1]
#        try:
        sax.parseString( d2, handler )
#        except SAXParseException:
#            self.status = -1
#            return 
        self.status = handler.status
        if self.status==0:
            self.nBeacons = handler.nBeacons
            self.simTime = handler.simTime
            #print "nBeacons", self.nBeacons

    def readSensors(self):
        data, (host,port) = self.sock.recvfrom(4096)
        d2 = data[:-1]

        #print "RECV : \"" + d2 +'"'
        parser = sax.make_parser()
        # Tell it what handler to use
        handler = StructureHandler()
        parser.setContentHandler( handler )

#        try:
        sax.parseString( d2, handler )
#        except SAXParseException:
#            status = -1
#            return 
        self.status = handler.status
        self.measures  = handler.measures
        
    def driveMotors(self, lPow, rPow):
        msg = '<Actions LeftMotor="'+str(lPow)+'" RightMotor="'+str(rPow)+'"/>'
        self.sock.sendto(msg.encode(),(self.host,self.port))

    def setReturningLed(self,val):
        msg = '<Actions LeftMotor="0.0" RightMotor="0.0" ReturningLed="'+ ("On" if val else "Off") +'"/>'
        self.sock.sendto(msg.encode(),(self.host,self.port))

    def setVisitingLed(self,val):
        msg = '<Actions LeftMotor="0.0" RightMotor="0.0" VisitingLed="'+ ("On" if val else "Off") +'"/>'
        self.sock.sendto(msg.encode(),(self.host,self.port))

    def finish(self):
        msg = '<Actions LeftMotor="0.0" RightMotor="0.0" EndLed="On"/>'
        self.sock.sendto(msg.encode(),(self.host,self.port))

    #my_status = lambda self : self.status


class CRobLinkAngs(CRobLink):


    def __init__ (self, robName, robId, angs, host):
        self.robName = robName
        self.robId = robId
        self.host = host
        self.angs = angs


        self.sock = socket.socket(socket.AF_INET, # Internet
                             socket.SOCK_DGRAM) # UDP
        
        msg = '<Robot Id="'+str(robId)+'" Name="'+robName+'">'
        for ir in range(NUM_IR_SENSORS):
            msg+='<IRSensor Id="'+str(ir)+'" Angle="'+str(angs[ir])+'" />'
        msg+='</Robot>'

        #print "msg ", msg
        
        self.sock.sendto(msg.encode(), (host, UDP_PORT))  # TODO condider host arg
        data, (host,self.port) = self.sock.recvfrom(1024)
        #print "received message:", data, " port ", self.port

        parser = sax.make_parser()
        
        # Tell it what handler to use
        handler = StructureHandler()
        parser.setContentHandler( handler )
        
        # Parse reply 

        d2 = data[:-1]
#        try:
        sax.parseString( d2, handler )
#        except SAXParseException:
#            self.status = -1
#            return 
        self.status = handler.status
        if self.status==0:
            self.nBeacons = handler.nBeacons
            self.simTime = handler.simTime
            #print "nBeacons", self.nBeacons

class CMeasures:

    def __init__ (self):
        self.compassReady=False
        self.compass=0.0; 
        self.irSensorReady=[False for i in range(NUM_IR_SENSORS)]
        self.irSensor=[0.0 for i in range(NUM_IR_SENSORS)]
        self.beaconReady = False   # TODO consider more than one beacon
        self.beacon = (False, 0.0)
        self.time = 0

        self.groundReady = False
        self.ground = -1
        self.collisionReady = False
        self.collision = False 
        self.start = False 
        self.stop = False 
        self.endLed = False
        self.returningLed = False
        self.visitingLed = False
        self.x = 0.0   
        self.y = 0.0   
        self.dir = 0.0

        self.scoreReady = False
        self.score = 100000
        self.arrivalTimeReady = False
        self.arrivalTime = 10000
        self.returningTimeReady = False
        self.returningTime = 10000
        self.collisionsReady = False
        self.collisions = 0
        self.gpsReady = False
        self.gpsDirReady = False


        self.hearMessage=''



from xml import sax

class StructureHandler(sax.ContentHandler):

    def __init__ (self):
        self.status = 0
        self.measures = CMeasures()

    def startElement(self, name, attrs):
        if name == "Reply":
            if "Status" not in attrs.keys():
                self.status = -1
                return
                
            if attrs["Status"]=="Ok":
                self.status = 0
                return
            self.status = -1
        elif name == "Parameters":
            self.nBeacons = attrs["NBeacons"]
            self.simTime = attrs["SimTime"]
        elif name=="Measures":
            self.measures.time = int(attrs["Time"])
        elif name=="Sensors":
            self.measures.compassReady         = ("Compass" in attrs.keys())
            if self.measures.compassReady:
                self.measures.compass =   float(attrs["Compass"])

            self.measures.collisionReady         = ("Collision" in attrs.keys())
            if self.measures.collisionReady:
                self.measures.collision = (attrs["Collision"] == "Yes")

            self.measures.groundReady         = ("Ground" in attrs.keys())
            if self.measures.groundReady:
                self.measures.ground =    int(attrs["Ground"])

        elif name == "IRSensor":
            id = int(attrs["Id"])
            #print "IRSensor id ", id
            if id < NUM_IR_SENSORS: 
                self.measures.irSensorReady[id] = True
                #print "IRSensor val ", attrs["Value"]
                self.measures.irSensor[id] = float(attrs["Value"])
            else: 
                self.status = -1
        elif name == "BeaconSensor":
            id = attrs["Id"]
            #if id<self.measures.beaconReady.len():
            if 1:
                self.measures.beaconReady=True
                if attrs["Value"] == "NotVisible":
                    #self.measures.beaconReady[id]=(False,0.0)
                    self.measures.beacon=(False,0.0)
                else:
                    #self.measures.beaconReady[id]=(True,attrs["Value"])
                    self.measures.beacon=(True,float(attrs["Value"]))
            else:
                self.status = -1
        elif name == "GPS":
            if "X" in attrs.keys():
                self.measures.gpsReady = True
                self.measures.x = float(attrs["X"])
                self.measures.y = float(attrs["Y"])
                if "Dir" in attrs.keys():
                     self.measures.gpsDirReady = True
                     self.measures.dir = float(attrs["Dir"])
                else:
                     self.measures.gpsDirReady = False
            else:
                self.measures.gpsReady = False
        elif name == "Leds":
            self.measures.endLed = (attrs["EndLed"] == "On")
            self.measures.returningLed = (attrs["ReturningLed"] == "On")
            self.measures.visitingLed = (attrs["VisitingLed"] == "On")
        elif name == "Buttons":
            self.measures.start = (attrs["Start"] == "On")
            self.measures.stop = (attrs["Stop"] == "On")
        elif name == "Score":
            self.measures.scoreReady         = ("Score" in attrs.keys())
            if self.measures.scoreReady:
                 self.measures.score=int(attrs["Score"])
            self.measures.arrivalTimeReady         = ("ArrivalTime" in attrs.keys())
            if self.measures.arrivalTimeReady:
                 self.measures.arrivalTime=int(attrs["ArrivalTime"])
            self.measures.returningTimeReady         = ("ReturningTime" in attrs.keys())
            if self.measures.returningTimeReady:
                 self.measures.returningTime=int(attrs["ReturningTime"])
            self.measures.collisionsReady         = ("Collisions" in attrs.keys())
            if self.measures.collisionsReady:
                 self.measures.collisions=int(attrs["Collisions"])
        elif name == "Message":
            self.hearFrom = int(attrs["From"])


#    def endElement(self, name):
        #print 'End of element:', name

